Stop Symbols from becoming Functions

cbm755 opened the pull request:
There are various problems caused by Symbols automatically becoming Functions.  That is:
```py
>>> x = Symbol('x')
>>> f = Symbol('f')
>>> f(x)
f(x)
>>> srepr(_)
"Function('f')(Symbol('x'))"
```
It seems this exists so that `S('f(x)')` works.   See #3539 for discussion.  I got here from #9057 which is no longer failing per se.  But there are various other bug fixes here that were found by making this change.

This PR consists of some fixes cherry-picked from Diofant, with appreciaton.

#### Release notes

<!-- BEGIN RELEASE NOTES -->
* core
  * Symbol `f` no longer automatically becomes a Function in `f(x)`.
<!-- END RELEASE NOTES -->

cbm755 committed:
pprint: add tests where beta/phi Symbols become Functions

This is #9057.  Just makes the error more obvious and hopefully
causes it to actually fail, in case the segfaulting is specific
to my system.

smichr commented:
Maybe you could give this a try:

```diff
diff --git a/sympy/core/function.py b/sympy/core/function.py
index 56bb4a4..7373363 100644
--- a/sympy/core/function.py
+++ b/sympy/core/function.py
@@ -441,8 +441,10 @@ def is_commutative(self):

     def _eval_evalf(self, prec):
         # Lookup mpmath function based on name
-        fname = self.func.__name__
         try:
+            if isinstance(self.func, UndefinedFunction):
+                raise AttributeError
+            fname = self.func.__name__
             if not hasattr(mpmath, fname):
                 from sympy.utilities.lambdify import MPMATH_TRANSLATIONS
                 fname = MPMATH_TRANSLATIONS[fname]
```

Commit:
Remove undocumented Symbol.__call__ helper

See
https://groups.google.com/forum/?fromgroups=#!topic/sympy/Gn0d3Udb69I

Closes sympy/sympy#3539

- - - -

Cherry picked with thanks from
https://github.com/diofant/diofant/commit/791f23e4bb215a76b1d7ec6ffbd7e3baf7abba5b

Fixes #3539.

Commit:
Use Function's instead of Symbol's in some places

Commit:
Correct general case in _linear_2eq_order1_type7

The answer was deadly wrong.  x0 and y0 - are functions of t
in general (particular solution), but C2 - is not (arbitrary constant).
Perhaps, the later was a typo (missed *).

Reference: http://eqworld.ipmnet.ru/en/solutions/sysode/sode0107.pdf

- - - -

Cherry-picked from Diofant commit 29c6e9fd440147c684da64410e5edfd58d48abeb,
with thanks.  Minor edits by Colin Macdonald when it didn't apply cleanly.

Commit:
Fix test_count_ops_visual(): ITE import clashed with ITE Symbol

Commit:
Fix several nasty parsing tests:

1) auto_symbol should emit Function('f'), not Symbol, for 'f(x)'

2) t = standard_transformations + (split_symbols, implicit_multiplication)
   str(parse_expr("ay(x+1)", transformations=t)) == 'a*y*(x + 1)'

3) t = standard_transformations + (convert_xor, function_exponentiation)
   str(parse_expr("exp^y(z)", transformations=t)) == '(E**z)**y'

cbm755 committed:
Merge remote-tracking branch 'origin/master' into 9057_workaround

cbm755 committed:
Symbols no longer become functions: add tests

cbm755 committed:
dsolve: add notes by some `symbols(function=True)` nonsense

I don't think this a meaningful thing; at best it signals something
semantic to the programmer.  Perhaps these should be `cls=Function`
(at least the ones in `_type10` should be, other two look ok.)

`_type10` is apparently untested as no tests fail, even though `u(t)`
should now be an exception when `u` is a Symbol.

cbm755 committed:
dsolve: clarify test

This is supposed to test what happens when the function is `C1(t)`
(that is, the usual constant name).  There is room for improvement
so link to a new issue.

cbm755 committed:
parser: fix imports and test that `S('f(x)')` still works

After the changes removing Symbols becoming Functions from
Issue #3539, I'm not sure this was tested anywhere.  Fix imports so
that Symbol is available (and use that instead of `sympy.Symbol`).

cbm755 commented on sympy/solvers/ode.py:
```diff
@@ -6934,11 +6934,12 @@ def _linear_2eq_order1_type7(x, y, t, r, eq):
         P = exp(Integral(r['d'],t))
         sol1 = C1*x0 + C2*x0*Integral(r['b']*F*P/x0**2, t)
-        sol2 = C1*y0 + C2(F*P/x0 + y0*Integral(r['b']*F*P/x0**2, t))
+        sol2 = C1*y0 + C2*(F*P/x0 + y0*Integral(r['b']*F*P/x0**2, t))
```
This is a nice example of the kind of "type errors" that creep in when Symbols can automatically become Functions.

cbm755 committed:
more fixes for `Symbol.__call__` removal

cbm755 committed:
still more fixes for `Symbol.__call__` removal

smichr commented:
`cls=Function` edits can be made to ode.py.

Also, how is E^y(x) parsed now? It was referred to in the unreviewed commit (f0e6ab7) you mentioned but I didn't see a test for it.

cbm755 commented:
Both on master and after this PR, it seems to do what I expect:
```py
>>> srepr(S('E^y(z)'))
"exp(Function('y')(Symbol('z')))"
```
I'll look into that a commit log a bit more.  At the very least, I will add the above as a test...

cbm755 commented:
BTW @smichr a long time ago you [suggested something above](https://github.com/sympy/sympy/pull/9064#issuecomment-76413809).  I've filed #15095 for that.

asmeurer commented:
I would double check that all the changes to the parser are tested. I recall there not being a ton of tests for the parser.

cbm755 committed:
parser: more tests for Symbols not being callable

Add tests to improve coverage for the parser changes [1] for Symbols
not being callable.  Minor docstring change.

[1] f0e6ab74dffc6bcc31cef7b0f99cd26e8ee9083c

cbm755 committed:
doslve: further cleanups to `function=True`

In two cases, it would be wrong to replace them with `cls=Function` because
the whole point is to sub in Symbols for functions for later manipulation.
I suppose they could stay as `function=True` but I think that's more likely
to confuse future maintainers than help them remember the semantics of these
variables.

The other case in ` _linear_2eq_order2_type10` is untested.  It looks like
they should be `cls=Function` so I made that change.  Filed #15097 asking for
a new test.

cbm755 committed:
parser: minor edit

cbm755 commented:
Ok, I went over f0e6ab7 and added a few tests.  It LGTM.

smichr commented on sympy/parsing/sympy_parser.py:
```diff
@@ -385,25 +389,28 @@ def _split_symbols(tokens, local_dict, global_dict):
-                            del result[-2:]
-                            result.extend([(NAME, "%s" % char),
-                                           (NAME, 'Symbol'), (OP, '(')])
+                            result.extend([(NAME, "%s" % char)])
```
This is the only line that was added that is not hit by tests of `parsing` according to my analysis. The following hits it:
```
>>> t=transformations = standard_transformations + (split_symbols, implicit_multiplication,)
>>> parse_expr('yx', local_dict={'x':w},transformations=t)
w*y
```

cbm755 committed:
parser: add test for coverage

I made `w` real just to make it clear it really takes it from the local
dict.

asmeurer commented:
Great to finally have this fixed.
